### PR TITLE
KAFKA-9719: Streams with EOS-beta should fail fast for older brokers

### PR DESCRIPTION
**Pull request by mjsax**
Call for review @abbccdda @guozhangwang 

System test run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3856/

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java [hunk 1]
```diff
@@ -86,29 +83,21 @@ public static void main(final String[] args) throws IOException {
         builder.<String, String>stream(SOURCE_TOPIC).groupByKey(Grouped.with(stringSerde, stringSerde))
             .count()
             .toStream()
-            .mapValues(new ValueMapper<Long, String>() {
-                @Override
-                public String apply(final Long value) {
-                    return value.toString();
-                }
-            })
+            .mapValues(Object::toString)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java:86]: Just code cleanup

File: streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java [hunk 2]
```diff
@@ -86,29 +83,21 @@ public static void main(final String[] args) throws IOException {
...
-                Throwable cause = e;
-                if (cause instanceof StreamsException) {
-                    while (cause.getCause() != null) {
-                        cause = cause.getCause();
-                    }
+        streams.setUncaughtExceptionHandler((t, e) -> {
+            Throwable cause = e;
+            if (cause instanceof StreamsException) {
+                while (cause.getCause() != null) {
+                    cause = cause.getCause();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java:94]: Just code cleanup

File: tests/kafkatest/services/streams.py [hunk 3]
```diff
@@ -396,11 +396,11 @@ def __init__(self, test_context, kafka):
 
 
 class StreamsBrokerCompatibilityService(StreamsTestBaseService):
-    def __init__(self, test_context, kafka, eosEnabled):
+    def __init__(self, test_context, kafka, processingMode):
```
Comment (mjsax) [inline tests/kafkatest/services/streams.py:399]: Switch from "boolean" flag to "string"

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py [hunk 4]
```diff
@@ -27,9 +27,10 @@
 class StreamsBrokerCompatibility(Test):
     """
     These tests validates that
-    - Streams 0.11+ w/ EOS fails fast for older brokers 0.10.2 and 0.10.1
-    - Streams 0.11+ w/o EOS works for older brokers 0.10.2 and 0.10.1
-    - Streams fails fast for 0.10.0 brokers
+    - Streams works for older brokers 0.11 (or newer)
+    - Streams w/ EOS-alpha works for older brokers 0.11 (or newer)
+    - Streams fails fast for older brokers 0.10.0, 0.10.2, and 0.10.1
+    - Streams w/ EOS-beta fails fast for older brokers 2.4 or older
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_compatibility_test.py:34]: Two tests overlap and one of them is redundant -- removing the redundant one as side cleanup

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py [hunk 5]
```diff
@@ -55,20 +56,28 @@ def setUp(self):
         self.zk.start()
 
    
-    @parametrize(broker_version=str(LATEST_0_10_2))
-    @parametrize(broker_version=str(LATEST_0_10_1))
-    def test_fail_fast_on_incompatible_brokers_if_eos_enabled(self, broker_version):
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_compatibility_test.py:60]: This is covered via `test_fail_fast_on_incompatible_brokers` that tests against 0.10.0, 0.10.1 and 0.10.2 already

In older KS versions, 0.10.1 and 0.10.2 where only not supported if EOS is used, but now those older brokers are not supported at all any longer, and thus, we can merge both tests.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java [hunk 6]
```diff
@@ -477,7 +476,29 @@ public void testGetGlobalConsumerConfigsWithGlobalConsumerOverridenPrefix() {
...
+    }
+
+    @Test
+    public void shouldNotSetInternalThrowOnFetchStableOffsetUnsupportedConfigToFalseInConsumerForEosDisabled() {
+        final Map<String, Object> consumerConfigs = streamsConfig.getMainConsumerConfigs(groupId, clientId, threadIdx);
+        assertThat(consumerConfigs.get("internal.throw.on.fetch.stable.offset.unsupported"), is(nullValue()));
+    }
+
+    @Test
+    public void shouldNotSetInternalThrowOnFetchStableOffsetUnsupportedConfigToFalseInConsumerForEosAlphsa() {
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java]: ForEosAlpha

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py [hunk 7]
```diff
@@ -55,20 +56,28 @@ def setUp(self):
...
+    @parametrize(broker_version=str(LATEST_2_4))
+    @parametrize(broker_version=str(LATEST_2_3))
+    @parametrize(broker_version=str(LATEST_2_2))
+    @parametrize(broker_version=str(LATEST_2_1))
+    @parametrize(broker_version=str(LATEST_2_0))
+    @parametrize(broker_version=str(LATEST_1_1))
+    @parametrize(broker_version=str(LATEST_1_0))
+    @parametrize(broker_version=str(LATEST_0_11_0))
+    def test_compatible_brokers_eos_disabled(self, broker_version):
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_compatibility_test.py:72]: This is an existing test (the diff is just weird)

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py [hunk 8]
```diff
@@ -79,11 +88,11 @@ def test_fail_fast_on_incompatible_brokers_if_eos_enabled(self, broker_version):
     @parametrize(broker_version=str(LATEST_1_1))
     @parametrize(broker_version=str(LATEST_1_0))
     @parametrize(broker_version=str(LATEST_0_11_0))
-    def test_compatible_brokers_eos_disabled(self, broker_version):
+    def test_compatible_brokers_eos_enabled(self, broker_version):
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_compatibility_test.py]: Add this new test as side cleanup -- not sure why we never tested EOS for older broker but only non-EOS?

File: tests/kafkatest/services/streams.py (hunk 3 above)
Comment (guozhangwang) [reply tests/kafkatest/services/streams.py:399]: Good call.

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py (hunk 4 above)
Comment (guozhangwang) [reply tests/kafkatest/tests/streams/streams_broker_compatibility_test.py:34]: +1

**Review by guozhangwang — COMMENTED** [review]
This is just a wild thought and not directly related to this PR: should we consider logging a different error message inside `StreamThread#runLoop` that if the thrown exception is `UnsupportedVersionException` and error message starts with `Broker unexpectedly doesn't support requireStable..` we can log a different message suggesting users that broker does not support `eos-beta` yet?

If you agrees we can update the `streams_broker_compatibility_test.py` waiting until a slight different message.

Otherwise, LGTM (feel free to merge).

**Comment by mjsax** [comment]
System tests failed with
```
[2020-03-27 01:38:29,836] ERROR stream-thread [kafka-streams-system-test-broker-compatibility-d80c5254-baf7-4935-990d-4ccf0c822f46-StreamThread-1] Encountered the following exception during processing and the thread is going to shut down:  (org.apache.kafka.streams.processor.internals.StreamThread)
org.apache.kafka.streams.errors.StreamsException: Error encountered trying to commit a transaction [stream-thread [kafka-streams-system-test-broker-compatibility-d80c5254-baf7-4935-990d-4ccf0c822f46-StreamThread-1] task [0_0]]
	at org.apache.kafka.streams.processor.internals.StreamsProducer.commitTransaction(StreamsProducer.java:180)
	at org.apache.kafka.streams.processor.internals.TaskManager.commitOffsetsOrTransaction(TaskManager.java:786)
	at org.apache.kafka.streams.processor.internals.TaskManager.commitInternal(TaskManager.java:752)
	at org.apache.kafka.streams.processor.internals.TaskManager.commitAll(TaskManager.java:732)
	at org.apache.kafka.streams.processor.internals.StreamThread.maybeCommit(StreamThread.java:814)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:687)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:543)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:502)
Caused by: org.apache.kafka.common.errors.UnsupportedVersionException: Attempted to write a non-default generationId at version 2
```

Created https://issues.apache.org/jira/browse/KAFKA-9776 for tracking and disabled those system tests for now.

Fixed failing unit tests, addressed review comments, and rebased to resolve merge conflict.

New system test run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3858/

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 9]
```diff
@@ -511,6 +512,21 @@ public void run() {
         } catch (final Exception e) {
             // we have caught all Kafka related exceptions, and other runtime exceptions
             // should be due to user application errors
+
+            if (e instanceof UnsupportedVersionException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:516]: @guozhangwang New special log message for old brokers.

File: streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java [hunk 10]
```diff
@@ -86,43 +83,46 @@ public static void main(final String[] args) throws IOException {
...
+        final boolean eosEnabled = processingMode.startsWith(StreamsConfig.EXACTLY_ONCE);
 
         System.out.println("send data");
         final Properties producerProperties = new Properties();
         producerProperties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
         producerProperties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
         producerProperties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        if (eosEnabled) {
+            producerProperties.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "broker-compatibility-producer-tx");
+        }
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java:114]: Smaller improvement: write input data with transactions, too.

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py [hunk 11]
```diff
@@ -44,7 +46,11 @@ def __init__(self, test_context):
                                   topics={
                                       self.input: {'partitions': 1, 'replication-factor': 1},
                                       self.output: {'partitions': 1, 'replication-factor': 1}
-                                  })
+                                  },
+                                  server_prop_overides=[
+                                      ["transaction.state.log.replication.factor", "1"],
+                                      ["transaction.state.log.min.isr", "1"]
+                                  ])
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_compatibility_test.py:53]: We have only one broker in this test and want to test with EOS

**Comment by mjsax** [comment]
Retest this please

**Comment by mjsax** [comment]
Not sure why, but Jenkins build does not run...

Retest this please.

**Comment by mjsax** [comment]
System tests passed.

Retest this please (come on Mr.Jenkins...)

**Comment by guozhangwang** [comment]
LGTM (we can mark the tests as ignored until the other PR is merged).

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
> LGTM (we can mark the tests as ignored until the other PR is merged).

Already ignored: https://github.com/apache/kafka/pull/8367/files#diff-7924cd3d80f77980ae421699e631e2ccR89

**Comment by mjsax** [comment]
`QueryableStateIntegrationTest` fails in both builds.

Retest this please.

**Comment by mjsax** [comment]
Java 8: `QueryableStateIntegrationTest`
Java 11: passed.

Retest this please.

**Comment by mjsax** [comment]
Java 8:
```
kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup
org.apache.kafka.streams.integration.QueryableStateIntegrationTest.concurrentAccesses
org.apache.kafka.streams.integration.QueryableStateIntegrationTest.concurrentAccesses
```

Java 11:
```
kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup
```

Retest this please.

**Pull request merged**